crates/red_knot_python_semantic/src/semantic_index/usedef.rs

```diff
@@ -0,0 +1,160 @@
+        &self.all_definitions[self.definitions_by_use[use_id].definitions.clone()]
+    }
+
+    pub(crate) fn use_may_be_unbound(&self, use_id: ScopedUseId) -> bool {
+        self.definitions_by_use[use_id].may_be_unbound
+    }
+
+    pub(crate) fn public_definitions(&self, symbol: ScopedSymbolId) -> &[Definition<'db>] {
+        &self.all_definitions[self.public_definitions[symbol].definitions.clone()]
+    }
+
+    pub(crate) fn public_may_be_unbound(&self, symbol: ScopedSymbolId) -> bool {
+        self.public_definitions[symbol].may_be_unbound
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct Definitions {
+    definitions: Range<usize>,
+    may_be_unbound: bool,
+}
+
+impl Default for Definitions {
+    fn default() -> Self {
+        Self {
+            definitions: 0..0,
+            may_be_unbound: true,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(super) struct FlowSnapshot {
+    definitions_by_symbol: IndexVec<ScopedSymbolId, Definitions>,
+}
+
+pub(super) struct UseDefMapBuilder<'db> {
+    all_definitions: Vec<Definition<'db>>,
+
+    definitions_by_use: IndexVec<ScopedUseId, Definitions>,
+
+    // builder state: currently visible definitions for each symbol
+    definitions_by_symbol: IndexVec<ScopedSymbolId, Definitions>,
+}
+
+impl<'db> UseDefMapBuilder<'db> {
+    pub(super) fn new() -> Self {
+        Self {
+            all_definitions: Vec::new(),
+            definitions_by_use: IndexVec::new(),
+            definitions_by_symbol: IndexVec::new(),
+        }
+    }
+
+    pub(super) fn add_symbol(&mut self, symbol: ScopedSymbolId) {
+        let new_symbol = self.definitions_by_symbol.push(Definitions::default());
+        debug_assert_eq!(symbol, new_symbol);
+    }
+
+    pub(super) fn record_def(&mut self, symbol: ScopedSymbolId, definition: Definition<'db>) {
+        let def_idx = self.all_definitions.len();
+        self.all_definitions.push(definition);
+        self.definitions_by_symbol[symbol] = Definitions {
+            #[allow(clippy::range_plus_one)]
+            definitions: def_idx..(def_idx + 1),
+            may_be_unbound: false,
+        };
+    }
+
+    pub(super) fn record_use(&mut self, symbol: ScopedSymbolId, use_id: ScopedUseId) {
+        let new_use = self
+            .definitions_by_use
+            .push(self.definitions_by_symbol[symbol].clone());
+        debug_assert_eq!(use_id, new_use);
+    }
+
+    pub(super) fn snapshot(&self) -> FlowSnapshot {
+        FlowSnapshot {
+            definitions_by_symbol: self.definitions_by_symbol.clone(),
+        }
+    }
+
+    pub(super) fn set(&mut self, state: &FlowSnapshot) {
+        let num_symbols = self.definitions_by_symbol.len();
+        self.definitions_by_symbol = state.definitions_by_symbol.clone();
+        self.definitions_by_symbol
+            .resize(num_symbols, Definitions::default());
+    }
+
+    pub(super) fn merge(&mut self, state: &FlowSnapshot) {
```

Comment:
The encapsulation of this method is unclear to me. It seems that we're extending the ranges of `definitions` without adding new items to `all_definitions`. What's guaranteeing that we can later dereference the definitions without having them pushed to `all_definitions`.

Reply:
`all_definitions` never shrinks or is mutated, it only grows. And `FlowSnapshot` can only come from a snapshot of some previous state (barring that it comes from some other builder/scope entirely, but this would be a bug regardless.) So both `state.definitions_by_symbol` and `self.definitions_by_symbol` must refer to indices that are already present in `self.all_definitions`.

Reply:
(Oh, and in some of these cases we do add new items to `all_definitions`, using `extend_from_within`.)

Reply:
Hopefully clearer with the docs in https://github.com/astral-sh/ruff/pull/12357